### PR TITLE
Update shadowsocks library

### PR DESCRIPTION
**Pull request by WendelHime**
Status: Ready for review!

- [x] Unit tests updated
- [x] E2E test

### CODE REVIEW

File: shadowsocks/local.go [hunk 1]
```diff
@@ -112,34 +83,43 @@ func ListenLocalTCPOptions(options *ListenerOptions) net.Listener {
...
-			log.Errorf("serving on %s: %v", l.Addr(), err)
+	authFunc := service.NewShadowsocksStreamAuthenticator(options.Ciphers, options.ReplayCache, options.ShadowsocksMetrics)
+	tcpHandler := service.NewTCPHandler(options.Listener.Addr().(*net.TCPAddr).Port, authFunc, options.ShadowsocksMetrics, timeout)
+	tcpHandler.SetTargetDialer(&LocalDialer{connections: l.connections})
+	handler := &Handler{TCPHandler: tcpHandler}
+
+	accept := func() (transport.StreamConn, error) {
+		switch l.wrapped.(type) {
+		case *tcpListenerAdapter, *net.TCPListener:
+			conn, err := l.wrapped.(*tcpListenerAdapter).AcceptTCP()
```
[the remark this answers is not in the document]
Comment (WendelHime) [reply shadowsocks/local.go:94]: Yeah, I'll remove that since by default we're always using `*tcpListenerAdapter`.

**Comment by WendelHime** [comment]
Thanks for the suggestions @garmr-ulfr ! I've added some changes to address those issues!

File: shadowsocks/dialer.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+type LocalDialer struct {
+	connections chan net.Conn
+}
+
+func (d *LocalDialer) DialStream(ctx context.Context, addr string) (transport.StreamConn, error) {
+	cliConn, ok := ctx.Value(ClientConnCtxKey{}).(transport.StreamConn)
+	if !ok {
+		return nil, errors.New("expected stream connection in context but received another type")
+	}
```
[the remark this answers is not in the document]
Comment (WendelHime) [reply shadowsocks/dialer.go:19]: Thanks for the suggestions! Applying them!

File: shadowsocks/dialer.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	c1, c2 := net.Pipe()
+	a := &tcpConnAdapter{c1}
+	b := &lfwd{
+		Conn:           c2,
+		remoteAddr:     cliConn.RemoteAddr(),
+		clientTCPConn:  cliConn,
+		upstreamTarget: addr,
+	}
+	d.connections <- b
```
Comment (hwh33) [inline shadowsocks/dialer.go:29]: Given how difficult it was for both of us to understand this area of the code, could you add a comment explaining what's going on here?  I think something like, "We treat the Lantern proxy as the upstream, effectively proxying from the client to the actual proxy code."  But feel free to use your own words of course!

File: shadowsocks/local.go [hunk 4]
```diff
@@ -112,34 +83,37 @@ func ListenLocalTCPOptions(options *ListenerOptions) net.Listener {
...
+		ctx = context.WithValue(ctx, ClientConnCtxKey{}, conn)
+		tcpHandler.Handle(ctx, conn)
+	}
+
+	go service.StreamServe(accept, handler)
 	return l
 }
 
+// ClientConnCtxKey is a context key being used to share the client connection
+type ClientConnCtxKey struct{}
```
Comment (hwh33) [inline shadowsocks/local.go:115]: Nit: this doesn't need to be exported, right?  We should refrain from exporting things unless necessary.

**Review by hwh33 — APPROVED** [review]
Looks mostly good to me - just left a couple of comments about small things.

**Review by garmr-ulfr — APPROVED** [review]
LGTM

**Pull request merged**